src/lavinmq/server.cr

```diff
@@ -254,53 +254,59 @@ module LavinMQ
       loop do
         break if closed?
         sleep Config.instance.stats_interval.milliseconds
-        update_stats_rates
-
-        interval = Config.instance.stats_interval.milliseconds.to_i
-        log_size = Config.instance.stats_log_size
-        rusage = System.resource_usage
-
-        {% for m in METRICS %}
-          until @{{m.id}}_log.size < log_size
-            @{{m.id}}_log.shift
+        @stats_loop_duration_seconds_total = Time.measure do
+          @stats_rates_duration_seconds = Time.measure do
+            update_stats_rates
           end
-          {% if m.id.ends_with? "_time" %}
-            {{m.id}} = rusage.{{m.id}}.total_milliseconds.to_i64
-            {{m.id}}_rate = (({{m.id}} - @{{m.id}}) / (interval * 1000)).round(2)
-          {% else %}
-            {{m.id}} = rusage.{{m.id}}.to_i64
-            {{m.id}}_rate = (({{m.id}} - @{{m.id}}) / interval).round(2)
-          {% end %}
-          @{{m.id}}_log.push {{m.id}}_rate
-          @{{m.id}} = {{m.id}}
-        {% end %}
-
-        until @rss_log.size < log_size
-          @rss_log.shift
-        end
-
-        rss = statm_rss(statm) || ps_rss
-        @rss = rss
-        @rss_log.push @rss
-
-        @mem_limit = cgroup_memory_max || System.physical_memory.to_i64
 
-        fs_stats = Filesystem.info(@data_dir)
-        until @disk_free_log.size < log_size
-          @disk_free_log.shift
-        end
-        disk_free = fs_stats.available.to_i64
-        @disk_free_log.push disk_free
-        @disk_free = disk_free
-
-        until @disk_total_log.size < log_size
-          @disk_total_log.shift
+          @system_metrics_duration_seconds = Time.measure do
```

Comment:
extract everything in this block into its own method?

Reply:
That's what I did, but [this](https://github.com/cloudamqp/lavinmq/compare/6a4e36913473340becf10debe2f4434cb138e32e..1f6b4721e248b4d9310ece34c6ea23429c72ee37) seems to have changed back some things.